proposals/interface-types/working-notes/scenarios/unicode_count.md

```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
-export adapters is to inline the export adapter into the import adapter. For example, we can combine the import and export adapters for `countCodes`:
+export adapters is to inline the export adapter into the import adapter. For
+example, we can combine the import and export adapters for `countCodes`:
 
 ```
 (func $countCodes
-  (param $ptr i32)(param $len i32) (result i32))
-  defer-scope
+  (param $ptr i32)(param $len i32) (result i32)
     local.get $ptr
     local.get $len
-    memory-to-string "memi"
-    let (local $str string)
+    string.from.utf8 "memi"
+    let (result i32)(local $str string)
       local.get $str
       string.size
-      call $malloc
-      deferred (i32)
-        let (local $tgt i32)
-          local.get $str
-          local.get $tgt
-          string-to-memory "memx"
-          call $countCodes_impl
-          i32-to-u32
-        end
-      finally
-        call $free
+      call-export "ex:malloc"
+      own (i32)
+        call-export "ex:free"
+      end
+      let (result u32) (local $tgt (owned i32))
+        local.get $str
+        local.get $tgt
+        owned.access
+        utf8.from.string "memx"
+        call-export "ex:countCodes_impl"
+        u32.from.i32
+      end
     end
-  end
-  u32-to-i32
+    i32.from.u32
 )
 ```
 
-The process of simplication involves taking pairs of lifting and lowering
-operators -- each typically originally coming from either the import or the
-export adapter -- and replacing them with core instructions that implement the
-combination.
+Our first task is to identify the correct positioning of the memory management
+functions -- specifically, where we can safely invoke the `call free` function
+to release the memory allocated for the string.
+
+One straightfoward technique for this is to add additional local variables to
+the adapter function -- that are used to hold the references collected by the
+`own` instruction itself.
 
-In the case of the `defer-scope` and `deferred` instructions we eliminate both
-and move the instructions within the `finally` sub-block to the end of the
-`defer-scope` block:
+At the same time, we can unwrap the various uses of `owned.access`, and invoke
+the wrapped code at the end of the combined adapter:
 
 ```
 (func $countCodes
-  (param $ptr i32)(param $len i32) (result i32))
-  local.get $ptr
-  local.get $len
-  memory-to-string "memi"
-    
-  let (local $str string)
-    local.get $str
-    string.size
-    call $malloc
-    local.tee $def_0
-    let (local $tgt i32)
+  (param $ptr i32)(param $len i32) (result i32)
+  (local $m1 i32)  ;; Used for owned block
+    local.get $ptr
+    local.get $len
+    string.from.utf8 "memi"
+    let (result i32)(local $str string)
       local.get $str
-      local.get $tgt
-      string-to-memory "memx"
-      call $countCodes_impl
-      i32-to-u32
+      string.size
+      call-export "ex:malloc"
+      local.tee $m1   ;; store captured references for later
```

Comment:
Thus far, we've avoided any `local.set`s or mutation of locals to ensure that adapter fusion can easily propagate defs to uses.

Reply:
I understand. But, setting is unavoidable here.
Also, remember that the tee instruction is a result of the fusion. It is not present in the original adapters.

Reply:
Oh hah, sorry, I hadn't noticed this was only in the fused result.  Nevermind :)